### PR TITLE
Bamboo Relay Connector

### PR DESCRIPTION
**Pull request by Arctek**
Added Bamboo Relay support.

### CODE REVIEW

**Comment by yzhang1994** [comment]
@Arctek Wow the PR looks great! Can't believe you were able to figure everything out with the little documentation we have.

@martinkou I've reviewed the CLI part and the market class and everything looks good so far. Please review when you have a chance. @Arctek mentioned that he has tested it on his end for a few days already, so let's try to merge it in and start testing it soon. Hopefully, we can get this in the next release!

File: wings/data_source/bamboo_relay_local_cluster_order_book_data_source.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
+#!/usr/bin/env python
```
Comment (martinkou) [inline wings/data_source/bamboo_relay_local_cluster_order_book_data_source.py]: Local cluster data sources were really only for CoinAlpha's internal use before hummingbot was released. I'd consider any local cluster order book data source to be legacy code by now. So there's no need to have this.

**Comment by martinkou** [comment]
Most of the code looks very similar to Radar Relay - which is understandable since Bamboo is another relay on the 0x protocol.

I think I'm ok with merging this after addressing my comment above. But afterwards we may need to refactor some of the code so we can reduce code duplication while retaining flexibility in case Bamboo and Radar would diverge some time later.

File: wings/tracker/bamboo_relay_order_book_tracker.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        self._saved_message_queues: Dict[str, Deque[BambooRelayOrderBookMessage]] = defaultdict(lambda: deque(maxlen=1000))
+        self._active_order_trackers: Dict[str, BambooRelayActiveOrderTracker] = defaultdict(BambooRelayActiveOrderTracker)
+        self._symbols: Optional[List[str]] = symbols
+
+    @property
+    def data_source(self) -> OrderBookTrackerDataSource:
+        if not self._data_source:
+            if self._data_source_type is OrderBookTrackerDataSourceType.EXCHANGE_API:
+                self._data_source = BambooRelayAPIOrderBookDataSource(symbols=self._symbols)
+            elif self._data_source_type is OrderBookTrackerDataSourceType.LOCAL_CLUSTER:
```
Comment (martinkou) [inline wings/tracker/bamboo_relay_order_book_tracker.py]: just raise an error if local cluster type is selected.

Comment (martinkou) [reply wings/tracker/bamboo_relay_order_book_tracker.py]: You can reference Coinbase Pro's order book tracker for how it's done.

**Comment by Arctek** [comment]
Yes the code is more or less the same as Radar's - as I largely implemented their API on my side as it saved having to do two sets of effort.

It's likely Bamboo will diverge from Radar in that it will support the 0x coordinator model in the coming weeks, so the flexibility that's added will probably need to support a two-mode operation, when it comes to signing transactions and cancelling orders.

File: wings/tracker/bamboo_relay_order_book_tracker.py (hunk 2 above)
Comment (Arctek) [reply wings/tracker/bamboo_relay_order_book_tracker.py]: Sure, just removed the local cluster option.

File: wings/data_source/bamboo_relay_local_cluster_order_book_data_source.py (hunk 1 above)
Comment (Arctek) [reply wings/data_source/bamboo_relay_local_cluster_order_book_data_source.py]: Makes sense - the Radar files still references this as well, so you might want to also take this out while you are at it.

**Pull request merged**